pkg/controller/garbagecollector/garbage_collector.go

```diff
@@ -0,0 +1,522 @@
+				patch, err := item.patchToUnblockOwnerReferences()
+				if err != nil {
+					return err
+				}
+				if _, err := gc.patchObject(item.identity, patch); err != nil {
+					return err
+				}
+				break
+			}
+		}
+		glog.V(2).Infof("at least one owner of object %s has FianlizerDeletingDependents, and the object itself has dependents, so it is going to be deleted with DeletePropagationForeground", item.identity)
+		return gc.deleteObject(item.identity, v1.DeletePropagationForeground)
+	default:
+		glog.V(2).Infof("delete object %s with DeletePropagationDefault", item.identity)
+		return gc.deleteObject(item.identity, v1.DeletePropagationDefault)
+	}
+}
+
+// process item that's waiting for its dependents to be deleted
+func (gc *GarbageCollector) processDeletingDependentsItem(item *node) error {
+	blockingDependents := item.blockingDependents()
+	if len(blockingDependents) == 0 {
+		glog.V(2).Infof("remove DeleteDependents finalizer for item %s", item.identity)
+		return gc.removeFinalizer(item, v1.FinalizerDeleteDependents)
+	} else {
+		for _, dep := range blockingDependents {
+			if !dep.deletingDependents {
+				glog.V(2).Infof("adding %s to attemptToDelete, because its owner %s is deletingDependents", dep.identity, item.identity)
+				gc.attemptToDelete.Add(dep)
+			}
+		}
+		return nil
+	}
+}
+
+// dependents are copies of pointers to the owner's dependents, they don't need to be locked.
+func (gc *GarbageCollector) orhpanDependents(owner objectReference, dependents []*node) error {
+	var failedDependents []objectReference
+	var errorsSlice []error
+	for _, dependent := range dependents {
+		// the dependent.identity.UID is used as precondition
+		patch := deleteOwnerRefPatch(dependent.identity.UID, owner.UID)
+		_, err := gc.patchObject(dependent.identity, patch)
+		// note that if the target ownerReference doesn't exist in the
+		// dependent, strategic merge patch will NOT return an error.
+		if err != nil && !errors.IsNotFound(err) {
+			errorsSlice = append(errorsSlice, fmt.Errorf("orphaning %s failed with %v", dependent.identity, err))
+		}
+	}
+	if len(failedDependents) != 0 {
+		return fmt.Errorf("failed to orphan dependents of owner %s, got errors: %s", owner, utilerrors.NewAggregate(errorsSlice).Error())
+	}
+	glog.V(5).Infof("successfully updated all dependents")
+	return nil
+}
+
+// attemptToOrphanWorker dequeues a node from the orphanQueue, then finds its
+// dependents based on the graph maintained by the GC, then removes it from the
+// OwnerReferences of its dependents, and finally updates the owner to remove
+// the "Orphan" finalizer. The node is added back into the orphanQueue if any of
+// these steps fail.
+func (gc *GarbageCollector) attemptToOrphanWorker() {
```

Comment:
Actually I want to merge orphan handling logic with the regular "attempToDelete" worker. Let's keep it as is for now?

(After the merge, the flow will be like the graph builder pushes the owner to the attempToDeleteQueue, mark the node as `waitingForDependentsOrphaned`, and the attempToDelete worker deletes the owner if it doesn't have anymore dependents, otherwise call the `orphanDependents` function.)